### DISCUSSION TITLE
textcat training error: Aborting -Errno 38- Function not implemented

### DISCUSSION THREAD

**Discussion by rvandomme** [explosion/spaCy#7598] [Help: Coding & Implementations] [answered]
Hello Spacy Team,
I am trying to migrate a project from spaCy V2 to V3.
It is a document classification project using textcat. It was working well on V2, with the same data.
I converted the datasets in V3 format and learned how to use the config file.
The training starts but gets interrupted.
Now I am stuck as I have no idea where the issue is.

I am in an Azure MLStudio environment.
I use spaCy v 3.0.5

Here is the error screen:

![image](https://user-images.githubusercontent.com/29764287/112769329-7140e980-8fee-11eb-8a93-5eebb5d5239c.png)

Here is the config file :

```
[paths]
train = "./data/train.spacy"
dev = "./data/valid.spacy"
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","textcat"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.textcat]
factory = "textcat"
threshold = 0.5

[components.textcat.model]
@architectures = "spacy.TextCatBOW.v1"
exclusive_classes = false
ngram_size = 1
no_output_layer = false
nO = null

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 2000
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
cats_score_desc = null
cats_micro_p = null
cats_micro_r = null
cats_micro_f = null
cats_macro_p = null
cats_macro_r = null
cats_macro_f = null
cats_macro_auc = null
cats_f_per_type = null
cats_macro_auc_per_type = null
cats_score = 1.0

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

I debugged the config:

![image](https://user-images.githubusercontent.com/29764287/112769438-f0362200-8fee-11eb-8744-18e045b5441e.png)

and the data

![image](https://user-images.githubusercontent.com/29764287/112769464-1d82d000-8fef-11eb-932f-f956275879da.png)

No idea what are misaligned tokens and how to fix it. If it is the issue ...
I don't understand how train data contains instances without mutually-exclusive classes. It has been verified and all observation has only one label. 

Thanks in advance for your help
Roger

**Comment by adrianeboyd** [comment, marked as answer]
Hmm, that error looks more like something is wrong with your v3 installation or environment. Can you check whether you can run this demo project:

```bash
spacy project clone pipelines/textcat_demo
cd textcat_demo
spacy project run all
```

For textcat, misaligned tokens aren't a problem. If you converted v2 JSON training data, there might be some minor changes in the tokenizer settings between v2 and v3 that lead to this warning, but you can ignore it.

It's possible that the warning about the train/dev mismatch is a bug in the `debug data` script, which is a bit cludge-y and not 100% updated for all aspects of v3. You can verify the training data manually with something like this:

```python
import spacy
from spacy.tokens import DocBin

nlp = spacy.blank("en")
db = DocBin().from_disk("train.spacy")
for doc in db.get_docs(nlp.vocab):
    assert list(doc.cats.values()).count(1.0) == 1
```

**Reply by rvandomme** [reply]
Thank you Adriane.
Spot on ...
I ran the demo project on my own computer environment and it works like a charm. But I don't have enough "juice" for the real project.
I ran it on the Azure MLStudio environment and got the same error.
I totally uninstalled spaCy and reinstalled it.
Got the same error. :-(
So something wrong with the environment
Would you have guidelines and recommendations for installation in Azure MLStudio ?
Thanks again
Roger

**Reply by adrianeboyd** [reply]
I'm afraid I don't really know. The closest link I could find was this: https://stackoverflow.com/questions/66276809/errno-38-function-not-implemented-occurred-when-installing-tensorflow-objec

**Reply by rvandomme** [reply]
Thank you Adriane
I will ask my IT support to get acquainted with spaCy dependencies and help me on this one.
I will let you know and post it here if I find a fix
Cheers
Roger

**Reply by rvandomme** [reply]
Found it !
Azure ML environment doesn't like the output path ./output.
Just ignore it in the CLI and it works.
Silly...
